LedgerSMB/Database/Change.pm

```diff
@@ -0,0 +1,280 @@
+}
+
+=head2 sha
+
+sha of sql content, stripped of comments and lines with only whitespace
+characters
+
+=cut
+
+sub sha {
+    my ($self) = @_;
+    return $self->{_sha} if $self->{_sha};
+    my $content = $self->content(1); # raw
+    my $normalized = join "\n",
+                     grep { /\S/ }
+                     map { my $string = $_; $string =~ s/--.*//; $string }
+                     split("\n", $content);
+    $self->{_sha} = Digest::SHA::sha512_base64($normalized);
+    return $self->{_sha};
+}
+
+
+
+=head2 content_wrapped($before, $after)
+
+Wrap a file with more statements in the same transaction.
+
+So you get something like:
+
+BEGIN;
+$before
+$self->content
+$after
+COMMIT;
+
+Useful for db updates so you can update version numbers or the like.
+
+=cut
+
+sub content_wrapped {
+    my ($self, $before, $after) = @_;
+    $before //= "";
+    $after //= "";
+    return $self->_wrap_transaction(
+        _wrap($self->content(1), $before, $after)
+    );
+}
+
+=head2 is_applied($dbh)
+
+Returns true if the current sha matches one that has been applied.
+
+=cut
+
+sub is_applied {
+    my ($self, $dbh) = @_;
+    my $sha = $self->sha;
+    my $sth = $dbh->prepare(
+        "SELECT * FROM db_patches WHERE sha = ?"
+    );
+    $sth->execute($sha);
+    my $retval = int $sth->rows;
+    $sth->finish;
+    return $retval;
+}
+
+=head2 run($dbh)
+
+Runs against the current dbh without tracking.
+
+=cut
+
+sub run {
+    my ($self, $dbh) = @_;
+    $dbh->do($self->content); # not raw
+}
+
+=head2 apply($dbh)
+
+Applies the current file to the db in the current dbh.
+
+=cut
+
+sub apply {
+    my ($self, $dbh) = @_;
+    my $need_commit = _need_commit($dbh);
+    my $before = "";
+    my $after;
+    my $sha = $dbh->quote($self->sha);
+    my $path = $dbh->quote($self->path);
+    if ($self->is_applied($dbh)){
+        $after = "
+              UPDATE db_patches
+                     SET last_updated = now()
+               WHERE sha = $sha;
+        ";
+    } else {
+        $after = "
+           INSERT INTO db_patches (sha, path, last_updated)
+           VALUES ($sha, $path, now());
+        ";
+    }
+    if ($self->{no_transactions}){
+        $dbh->do($after);
+        $after = "";
+        $dbh->commit if $need_commit;
+    }
+    $dbh->do($self->content_wrapped($before, $after));
+    $dbh->commit if $need_commit;
+    warn "$dbh->state: $dbh->errstr";
+    $dbh->prepare("
+            INSERT INTO db_patch_log(when_applied, path, sha, sqlstate, error)
+            VALUES(now(), $path, $sha, ?, ?)
+    ")->execute($dbh->state, $dbh->errstr);
+    $dbh->commit if $need_commit;
+}
+
+sub _need_commit{
+    my ($dbh) = @_;
+    1; # todo, detect existing transactions and autocommit status
+}
+=head1 Package Functions
+
+=head2 init($dbh)
+
+Initializes the tracking system
+
+=cut
+
+sub init {
+    my ($dbh) = @_;
+    return 0 unless needs_init($dbh);
+    $dbh->do("
+    CREATE TABLE db_patch_log (
```

Comment:
What happens on failure? If we depend on RaiseError to be true, then we should document here and where we receive the handle.